### PR TITLE
Scripts: Add TypeScript for builds and tests

### PR DESCRIPTION
**Pull request by gziolo**
## Description

Inspired by https://github.com/mkaz/tsblock from @mkaz:

> An example repository with the basic plumbing in place to build a WordPress block using Typescript.

I thought that it would be great better TypeScript support directly to the WordPress packages. This PR improves configuration to make the build work with the example block.

## How has this been tested?

1. I cloned the repository https://github.com/mkaz/tsblock inside the Gutenberg folder.
2. I removed the existing webpack config override.
3. I executed `../node_modules/.bin/wp-scripts build` to confirm that the build works as expected.
4. I executed `../node_modules/.bin/wp-scripts start` to confirm that the build works as expected.

## Screenshots 

Screen Shot 2021-11-05 at 15 53 05

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

File: packages/scripts/config/webpack.config.js [hunk 1]
```diff
@@ -115,6 +112,7 @@ const config = {
 		alias: {
 			'lodash-es': 'lodash',
 		},
+		extensions: [ '.ts', '.tsx', '...' ],
```
Comment (gziolo) [inline packages/scripts/config/webpack.config.js:115]: I found the trick with `...` in the webpck docs. It's a way to include all default extensions.

**Comment by gziolo** [comment]
I still need to update the documentation to reflect new capabilities added but code-wise it should be fully functional 😄

**Comment by gziolo** [comment]
> I still need to update the documentation to reflect new capabilities added but code-wise it should be fully functional 😄

It's ready for review.

File: packages/scripts/config/webpack.config.js (hunk 1 above)
Comment (mkaz) [reply packages/scripts/config/webpack.config.js:115]: Nice!

**Review by mkaz — APPROVED** [review]
Great update 👍 This tests well and worked fine. 

One question, why doesn't webpack need `ts-loader`? Reading the documentation and how I implemented, I mapped the `.ts` and `.tsx` extensions to `ts-loader`. I don't see that being done, but it still works. 🤔

**Comment by gziolo** [comment]
> One question, why doesn't webpack need `ts-loader`? Reading the documentation and how I implemented, I mapped the `.ts` and `.tsx` extensions to `ts-loader`. I don't see that being done, but it still works. 🤔

Both work. You can either use the TypeScript engine directly or Babel that also supports TypeScript syntax. I figured out it’s better to follow the way we handle TypeScript in Gutenberg with Babel.

https://babeljs.io/docs/en/babel-preset-typescript

https://github.com/WordPress/gutenberg/blob/161ee18541d030f611e5efb64453db37d804cc83/packages/babel-preset-default/index.js#L70

**Comment by mkaz** [comment]
> Babel that also supports TypeScript syntax

Ah right. Babel. I was so focused on webpack and its config. I didn't look at the babel-preset config.
Thanks! 👍

**Comment by overclokk** [comment]
Is this already available? Because I'm using the latest wp-script 19.2.2 but does not compile the ts[x] files.

**Comment by gziolo** [comment]
Unfortunately, we are blocked by the release process for the next major WordPress version. It won’t be published to npm until January. We need to rethink the npm publishing process now that we got blocked for more than two months ☹️

**Comment by overclokk** [comment]
Thank you for your answer, for now I will try to use the script from @mkaz , it's a pity to wait untill jannuary ☹️

**Comment by overclokk** [comment]
So, for now I just copied the webpack config file in my directory and it seems to work.
I also copied the two files for jest config and changed in all file the path for `require( '../utils' )` to `require( '@wordpress/scripts/utils' )` and also changed other paths to point to the right directory to avoid arror on run, for example in `jest-unit.config.js` I updated the path with `@wordpress/scripts/config/` to point to the right file.
I also copied the all jest-preset-default directory and point to this copied directory instead of the `@wordpress` package.

As soon as it will be released I will revert to the default version.

**Pull request merged**